### PR TITLE
Module for HTMX SSE extension

### PR DESCRIPTION
**Pull request by h0p3zZ**
* Added SSE extension support in seperate library
* Example code vor Vapor and Hummingbird

### CODE REVIEW

File: Sources/ElementaryHxSSE/HTMLAttribute+HxSSE.swift [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import Elementary
+import ElementaryHTMX
+
+public extension HTMLAttribute.hx where Tag: HTMLTrait.Attributes.Global {
+    /// A namespace for the HTMX SSE-extension.
+    /// See the [htmx-sse reference](https://github.com/bigskysoftware/htmx-extensions/blob/main/src/sse/README.md) for more information.
+    enum sse {}
+}
```
Comment (sliemeobn) [inline Sources/ElementaryHxSSE/HTMLAttribute+HxSSE.swift]: in order to keep the spelling close to plain HTML I prefer to have the `.sse` enum on the HTMLAttribute directly.
this way we can spell it as `div(.hx.ext(.sse), .sse.connect("my-url")) {}` and stick to the general pattern of elementary.

File: Sources/ElementaryHTMX/HTMLAttributeValue+HTMX.swift [hunk 2]
```diff
@@ -183,6 +183,22 @@ public extension HTMLAttributeValue.HTMX.EventTrigger {
...
+        public init(rawValue: String) {
+            self.rawValue = rawValue
+        }
+
+        public init(stringLiteral value: String) {
+            rawValue = value
+        }
+
+        public static var sse: Self { "sse" }
+        // add other extensions here
```
Comment (sliemeobn) [inline Sources/ElementaryHTMX/HTMLAttributeValue+HTMX.swift]: the type `Extension` is in the right place, but I prefer the `sse` static initializer in the SSE module as an extension.

File: Package.swift [hunk 3]
```diff
@@ -21,6 +21,10 @@ let package = Package(
             name: "ElementaryHTMX",
             targets: ["ElementaryHTMX"]
         ),
+        .library(
+            name: "ElementaryHxSSE",
+            targets: ["ElementaryHxSSE"]
+        ),
```
Comment (sliemeobn) [inline Package.swift]: after going back and forth I think I the name should be `ElementaryHTMXSSE`

File: Sources/ElementaryHxSSE/HTMLAttribute+HxSSE.swift [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        .init(name: "sse-swap", value: eventName)
+    }
+
+    static func close(_ eventName: String) -> HTMLAttribute {
+        .init(name: "sse-close", value: eventName)
+    }
+
+    static func trigger(_ eventName: String) -> HTMLAttribute {
+        .init(name: "hx-trigger", value: "sse:\(eventName)")
+    }
```
Comment (sliemeobn) [inline Sources/ElementaryHxSSE/HTMLAttribute+HxSSE.swift]: re-defining `trigger` here feels wrong. it is both in the wrong scope (in html there is "sse-trigger" but "hx-trigger").

I think we should simply add an extension to the `HTMLAttributeValue.HTMX.TriggerEvent` type roughly like so
```swift
extensions HTMLAttributeValue.HTMX.TriggerEvent {
   public static func sse(_ eventName: String) -> Self { .init(rawValue: "sse:\(eventName)") }
}
```

File: Examples/Vapor/VaporSSEDemo/Sources/App/Routes.swift [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+                } catch {
+                    req.logger.error("Stream Error: \(error)")
+                    // _ = writer.write(.error(error)) // writing error would crash the app
+                }
+                _ = writer.write(.end)
+            }
+        })
+        let res = Response(status: .ok, body: body)
+        res.headers.replaceOrAdd(name: "Content-Type", value: "text/event-stream")
+        return res
```
Comment (sliemeobn) [inline Examples/Vapor/VaporSSEDemo/Sources/App/Routes.swift]: ideally there should be no unstructured Task here

**Review by sliemeobn — CHANGES_REQUESTED** [review]
thanks for the contribution.

I really think the addition of a little sse-based live time view does not justify adding (and maintaining) two more example projects. Also, moving the existing examples will break existing links to them (in forum posts or similar)

Please rework the examples to keep the existing folder structure and simply add the live-time view snippets in there. It will be nice to have a little annotation on the page highlighting that the time is actually the "server time" or similar.

File: Sources/ElementaryHTMXSSE/HTMLAttribute+HTMXSSE.swift [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+import ElementaryHTMX
+
+public extension HTMLAttribute where Tag: HTMLTrait.Attributes.Global {
+    /// A namespace for the HTMX SSE-extension.
+    /// See the [htmx-sse reference](https://github.com/bigskysoftware/htmx-extensions/blob/main/src/sse/README.md) for more information.
+    enum sse {}
+}
+
+// TODO: implement API for event listeners for HTMX-SSE-events
+// https://github.com/bigskysoftware/htmx-extensions/blob/main/src/sse/README.md#listening-to-events-dispatched-by-this-extension
```
Comment (sliemeobn) [inline Sources/ElementaryHTMXSSE/HTMLAttribute+HTMXSSE.swift]: is this comment relevant?

File: Sources/ElementaryHTMXSSE/HTMLAttribute+HTMXSSE.swift [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        .init(name: "sse-connect", value: url)
+    }
+
+    static func swap(_ eventName: String) -> HTMLAttribute {
+        .init(name: "sse-swap", value: eventName)
+    }
+
+    static func close(_ eventName: String) -> HTMLAttribute {
+        .init(name: "sse-close", value: eventName)
+    }
```
Comment (sliemeobn) [inline Sources/ElementaryHTMXSSE/HTMLAttribute+HTMXSSE.swift:21]: does this even exist?

File: Examples/VaporDemo/Sources/App/Views.swift [hunk 8]
```diff
@@ -1,17 +1,23 @@
 import Elementary
 import ElementaryHTMX
+import ElementaryHTMXSSE
+import Foundation
 
 struct MainPage: HTMLDocument {
-    var title: String { "Vapor + Elementary + HTMX" }
+    var title: String { "Vapor + Elementary + HTMX + HTMX SSE" }
```
Comment (sliemeobn) [inline Examples/VaporDemo/Sources/App/Views.swift]: I think we should leave it at "Vapor + Elementary + HTMX"
sse is just a detail

File: Examples/VaporDemo/Sources/App/Routes.swift [hunk 9]
```diff
@@ -15,4 +16,20 @@ func addRoutes(to app: Application) {
...
+        })
+        let res = Response(status: .ok, body: body)
+        res.headers.replaceOrAdd(name: "Content-Type", value: "text/event-stream")
+        return res
+    }
+}
+
+struct Event: Encodable {
+    var event: String
+    var data: String
```
Comment (sliemeobn) [inline Examples/VaporDemo/Sources/App/Routes.swift]: is this used?

File: Examples/VaporDemo/Sources/App/Views.swift [hunk 10]
```diff
@@ -1,17 +1,23 @@
...
     var head: some HTML {
         meta(.charset(.utf8))
         script(.src("/htmx.min.js")) {}
+        script(.src("/htmxsse.min.js")) {}
     }
 
     var body: some HTML {
         header {
-            h2 { "Vapor + Elementary + HTMX" }
+            h2 { "Vapor + Elementary + HTMX + HTMX SSE" }
```
Comment (sliemeobn) [inline Examples/VaporDemo/Sources/App/Views.swift]: same here, "Vapor + Elementary + HTMX" is fine

File: Examples/HummingbirdDemo/Sources/App/ElementaryHTMXSSE+Hummingbird.swift
Comment (sliemeobn) [inline Examples/HummingbirdDemo/Sources/App/ElementaryHTMXSSE+Hummingbird.swift:1]: I think the file name is off, it should be more like "Response+Stream" or something

File: Examples/HummingbirdDemo/Sources/App/Routes.swift [hunk 11]
```diff
@@ -44,3 +53,8 @@ func addRoutes(to router: Router<some RequestContext>) {
 struct AddItemRequest: Decodable {
     var item: String
 }
+
+struct Event: Encodable {
+    var event: String
+    var data: String
+}
```
Comment (sliemeobn) [inline Examples/HummingbirdDemo/Sources/App/Routes.swift]: is this used?

File: Examples/HummingbirdDemo/Sources/App/Views.swift [hunk 12]
```diff
@@ -1,20 +1,27 @@
...
     }
 
     var body: some HTML {
         header(.class("container")) {
-            h2 { "Hummingbird + Elementary + HTMX Demo" }
+            h2 { "Hummingbird + Elementary + HTMX + HTMX SSE Demo" }
+            div(.hx.ext(.sse), .sse.connect("/time"), .sse.swap("time")) {
+                h4 {
+                    "Server Time: \(Date())"
+                }
```
Comment (sliemeobn) [inline Examples/HummingbirdDemo/Sources/App/Views.swift]: Should use `TimeHeading` I guess, right?

**Review by sliemeobn — CHANGES_REQUESTED** [review]
almost there

File: Sources/ElementaryHTMXSSE/HTMLAttribute+HTMXSSE.swift (hunk 7 above)
Comment (h0p3zZ) [reply Sources/ElementaryHTMXSSE/HTMLAttribute+HTMXSSE.swift:21]: https://www.npmjs.com/package/htmx-ext-sse yes - at least the doc says so

**Review by sliemeobn — APPROVED** [review]
lgtm (tiny example fixes are needed, but I'll do them separately)

**Pull request merged**